core/iwasm/aot/aot_loader.c

```diff
@@ -2532,6 +2535,68 @@ load_init_data_section(const uint8 *buf, const uint8 *buf_end,
     return false;
 }
 
+static bool
+merge_data_and_text(const uint8 **buf, const uint8 **buf_end, AOTModule *module,
+                    char *error_buf, uint32 error_buf_size)
+{
+    uint8 *old_buf = (uint8 *)*buf;
+    uint8 *old_end = (uint8 *)*buf_end;
+    uint32 code_size = (uint32)(uint64)(old_end - old_buf);
+    uint32 k_page_size = os_getpagesize();
+    bool need_merge = false;
+    uint32 total = 0, i;
+    uint8 *sections;
+    if (code_size == 0) {
+        return true;
+    }
+    /* calc total memory needed */
+    total += (code_size + k_page_size - 1) & ~(k_page_size - 1);
+    for (i = 0; i < module->data_section_count; ++i) {
+        AOTObjectDataSection *data_section = module->data_sections + i;
+        total += (data_section->size + k_page_size - 1) & ~(k_page_size - 1);
+        uint64 diff =
+            old_end < data_section->data
+                ? (uint64)(data_section->data + data_section->size - old_end)
+                : (uint64)(old_buf - data_section->data);
+        if (diff > ((int64)4 * BH_GB)) {
+            need_merge = true;
+        }
+    }
+    if (need_merge) {
+        int map_prot = MMAP_PROT_READ | MMAP_PROT_WRITE | MMAP_PROT_EXEC;
+#if defined(BUILD_TARGET_X86_64) || defined(BUILD_TARGET_AMD_64) \
+    || defined(BUILD_TARGET_RISCV64_LP64D)                       \
+    || defined(BUILD_TARGET_RISCV64_LP64)
+        /* aot code and data in x86_64 must be in range 0 to 2G due to
+         relocation for R_X86_64_32/32S/PC32 */
+        int map_flags = MMAP_MAP_32BIT;
+#else
+        int map_flags = MMAP_MAP_NONE;
+#endif
+        sections = module->merged_sections =
+            os_mmap(NULL, total, map_prot, map_flags, os_get_invalid_handle());
+        if (!sections) {
+            return false;
+        }
+        /* data first */
```

Comment:
No.
I just put data first because it is parsed in order `AOT_SECTION_TYPE_INIT_DATA` then `AOT_SECTION_TYPE_TEXT`.
Compiler normally put .data section after .text section.

Reply:
new commit change memory order as compilers do.